feature/196 사용자 닉네임 설정에 제한사항을 건다.

YejiAhn opened the pull request:
Resolves #196 
- 닉네임 길이 관련 제한사항을 백엔드, 프론트엔드 각각에서 걸어주었고
- 숫자/영어/한글 로만 이루어진 닉네임을 설정해야한다는 제약사항을 프론트에 걸어주고
- 관련된 css들 손보았습니다.

리뷰는 월요일까지 부탁드려요 🙌

YejiAhn committed:
feature/196 feat: 백엔드 코드에 Member 닉네임 글자 수 제한 사항(최대 15자) 설정

YejiAhn committed:
feature/196 feat: 프론트엔드 코드에 Member 닉네임 글자 수 제한 사항(최대 15자) 설정

  - 최대 글자 수(15자) 설정
  - 글자 수 세어주는 counter 기능 추가

YejiAhn committed:
feature/196 feat: 프론트엔드 닉네임 관련 css들 리팩토링

 - 사이드바에서 닉네임 11글자까지만 보여주게 한다.
 - 글/댓글의 프로필사진과 닉네임 간 간격 미세 조정한다.

YejiAhn committed:
feature/196 feat: 닉네임에 한글/숫자/영어만 쓰게 하는 제한사항 설정

 - :rules 속성으로 밑에 위반 메세지 보여주기
 - 제한사항을 어겼을 시 저장 버튼 비활성화

YejiAhn committed:
feature/196 refactor: 포매터 적용

hwookim commented on front/src/components/member/MemberSidebar.vue:
```diff
@@ -1,16 +1,16 @@
       </v-icon>
 
-      <div class="ma-5 mt-10 mb-5" @click="login">
+      <div class="ma-3 mt-10 mb-4" @click="login">
```
```suggestion
      <div class="mx-3 mt-10 mb-4" @click="login">
```

생각해보니 이거랑 똑같네요.

hwookim commented on front/src/components/member/MemberSidebar.vue:
```diff
@@ -83,6 +84,18 @@ export default {
+      if (Array.from(this.member.nickname).length > NICKNAME_LENGTH) {
+        return "...";
+      }
+      return "";
```
```suggestion
      return this.member.nickname.length > NICKNAME_LENGTH ? "..." : "";
```

특수문자, 이모티콘을 제외했으니 Array로 변환하는 과정이 필요하지 않을 것 같아요.😄

hwookim commented on front/src/components/member/MemberUpdateModal.vue:
```diff
@@ -66,8 +79,11 @@ export default {
+    nicknameUpdateDisabled() {
+      return (
+        this.member.nickname === this.newNickname ||
+        !this.newNickname.match(NICKNAME_REGEX)
```
같은 룰이 `this.rules`에서도 쓰이고 있는데요.
어떻게 하면 두 규칙이 한곳에서 관리될 수 있을까요?🤔

hwookim commented on src/test/java/com/grasshouse/dorandoran/member/repository/MemberRepositoryTest.java:
```diff
@@ -0,0 +1,53 @@
+        Member member = Member.builder().nickname("일이삼사오육칠팔구십일이삼사오육").build();
+        Member member2 = Member.builder().nickname("ottffssentottffs").build();
+        Member member3 = Member.builder().nickname("1234567890123456").build();
+        assertThatThrownBy(() -> {
+            memberRepository.save(member);
+            memberRepository.save(member2);
+            memberRepository.save(member3);
+        }).isInstanceOf(ConstraintViolationException.class);
```
~~이게 테스트가 될까요?
첫번째 save에서 에러가 발생하고 두번째는 실행이 안되지 않나요? 🤔~~
여러 닉네임을 넣어봐야 한다면 `@ParameterziedTest`를 사용해보는 건 어떨까요?
위의 정상적인 테스트도 마찬가지로요. 😄

chws replied:
추가적으로 첫번째 멤버에 가능한 닉네임을 넣어도 테스트가 통과하네요! 이 부분 수정해주세요😊

hwookim replied:
assertAll은 내부의 테스트 중 한가지가 실패하더라도 남은 테스트를 모드 검증하더군요!
잘못된 지식으로 리뷰를 남겨 죄송합니다.😢

hwookim commented on src/main/java/com/grasshouse/dorandoran/member/domain/Member.java:
```diff
@@ -34,6 +35,7 @@
     @GeneratedValue(strategy = GenerationType.IDENTITY)
     private Long id;
 
+    @Length(max = 15, message = "닉네임은 15자를 초과할 수 없습니다.")
```
Regex에 대한 예외처리는 따로 없을까요?

hwookim commented on front/src/components/member/MemberUpdateModal.vue:
```diff
@@ -6,7 +6,10 @@
+          maxlength="15"
+          :counter="15"
           v-model="newNickname"
+          :rules="[rules.violated, rules.unchanged]"
```
```suggestion
          :rules="Object.values(rules)"
```
기존의 방식은 룰이 추가될 때마다 같이 수정해줘야 할거예요.
대신 이렇게 표현할 수도 있겠네요 😃

hwookim commented on front/src/components/member/MemberUpdateModal.vue:
```diff
@@ -35,12 +38,22 @@
+const NICKNAME_REGEX = "^[a-zA-Zㄱ-ㅎㅏ-ㅣ가-힣0-9]+$";
+
 export default {
   name: "PostCreateModal",
```
띠용👀
이름이 이상했었군요...!

hwookim commented on front/src/components/member/MemberUpdateModal.vue:
```diff
@@ -35,12 +38,22 @@
     return {
       rendered: false,
       newNickname: "",
+      rules: {
+        violated: (newVal) =>
+          !!newVal.match(NICKNAME_REGEX) ||
+          "닉네임은 숫자/한글/영어로 이루어져야 합니다.",
+        unchanged: (newNickname) =>
+          this.member.nickname !== newNickname ||
+          "새로운 닉네임을 입력해주세요.",
+      },
```
```javascript

const NICKNAME_RULES = {
  VIOLATED: {
    rule: (newVal) => !!newVal.match(NICKNAME_REGEX),
    message: "닉네임은 숫자/한글/영어로 이루어져야 합니다.",
  },
  UNCHANGED: {
    rule: (newVal, oldVal) => newVal !== oldVal,
    message: "새로운 닉네임을 입력해주세요.",
  },
};

export default {
  name: "MemberUpdateModal",
  ...,
  computed: {
    rules() {
      return Object.values(NICKNAME_RULES).map(
        (value) =>
          value.rule(this.newNickname, this.member.nickname) || value.message,
      );
    },
  },
  ...
  methods: {
    nicknameUpdateDisabled() {
      const validation = Object.values(NICKNAME_RULES).filter(
        ({ rule }) => !rule(this.newNickname, this.member.nickname),
      );
      return validation.length !== 0;
    },
  },
```
이런 방식은 좀 별로려나요...?
앞으로 룰의 추가가 좀 더 쉬워지겠지만, 지금 당장 코드가 좀 복잡하네요ㅠㅠ
좀 더 고민해볼게요!

YejiAhn replied:
훨씬 관리되는 느낌이 나고 좋은데요! 💯 반영해볼게요!

hwookim commented on front/src/components/member/MemberUpdateModal.vue:
```diff
@@ -66,8 +79,11 @@ export default {
     },
-    nicknameNotChanged() {
-      return this.member.nickname === this.newNickname;
+    nicknameUpdateDisabled() {
```
지금은 정보수정이 닉네임 뿐이지만, 후에는 어떻게 될 지 모르겠군요.
등록버튼의 대상이 닉네임에 한정되지 않을 수 있어요.
닉네임이라고 명시하지 않아도 괜찮을 것 같아요.

다른 동작을 하는 게 아니라 결과값만 도출해내는 메소드라, 메소드 보다는 computed가 더 어울려보여요.
`isUpdatePossoble()`같은 네이밍으로 바꾸고 computed로 옮기는 건 어때요?

hwookim reviewed (changes requested):
밤 늦게까지 고생하셨습니다!!👍
되게 간단하게 생각했는데, 생각보다 구현하려니 복잡하군요.
textfield rules의 결과값에 바로 btn disable을 바인딩할 수 있으면 좋을텐데요...
쉽지 않네요 😢

몇가지 리뷰 남겼으니 확인 부탁드려요 ㅎㅎ🥰

yebink commented on front/src/components/member/MemberUpdateModal.vue:
```diff
@@ -35,12 +38,22 @@
+        violated: (newVal) =>
+          !!newVal.match(NICKNAME_REGEX) ||
+          "닉네임은 숫자/한글/영어로 이루어져야 합니다.",
+        unchanged: (newNickname) =>
```
`rules.unchanged`에 위반되면 아래와 같은 설명이 나오는데요!
![image](https://user-images.githubusercontent.com/19922698/93001550-bd54a100-f56a-11ea-823b-bea44514fe88.png)
설명이 상당히 중복된다는 느낌이 들어요.

+사실 해당 rule이 필요한지 의문입니다... 불필요한 요청을 막기 위해 만들어뒀지만 사용성을 해치는 느낌..?


또한 최초 로그인 시의 닉네임을 바꾸지 않으면 매번 MemberUpdateModal을 보게 되는데, 이 부분도 수정이 필요해보입니다. (죄송합니다 지난번에 제가 수정했어야 하는 부분인데 ㅠㅠ)

확인 결과, update api는 쏴지는데 jpa 내부적으로 같은 값이기 때문에 update 쿼리가 안 날라가고, 따라서 @LastModifiedDate도 작동하지 않고 있네요.
Member.update()에서 수동으로 날짜를 update해주니까 update 쿼리도 날라가고, 초기 닉네임을 사용하지 못하는 문제도 해결할 수 있을 것 같아요!

![image](https://user-images.githubusercontent.com/19922698/93001653-70bd9580-f56b-11ea-8ac4-c26cf697fa84.png)

YejiAhn replied:
맞죠 저도 짜면서도 메세지가 중복된다는 생각이 들었는데.. 
다른 몇몇 서비스들은 에러 메세지 없이 확인버튼만 비활성화로만 막아놓는지 하는 방법을 쓰더라구요?
생각해보면 같은 닉네임으로 수정하는데 updatedAt 시간을 바꿔줘야하냐는 의문이 들기도 하구요 🤔 
다음주 첫 회의 때 안건으로 말해보도록 할게요 짚어주셔서 감사합니다 👍 
 
그리고 최초 회원가입 시 닉네임 이슈가 #186 이슈랑 지금 이 이슈로 분리 되어서요,
해당 이슈 새로 파서 진행해야될 것 같아요. 다음 주에 진행하도록 하죠 🙂

yebink commented on front/src/components/member/MemberUpdateModal.vue:
```diff
@@ -35,12 +38,22 @@
     return {
       rendered: false,
       newNickname: "",
+      rules: {
```
![image](https://user-images.githubusercontent.com/19922698/93002643-5d162d00-f573-11ea-8aaa-66d6691bf204.png)
15자 이상이 되었을 경우도 메시지를 표시해주면 좋을 것 같아요!

LTTTTTE replied:
맥 한글입력 문제같네요 이건

YejiAhn replied:
맞습니다 영어나 숫자는 15자 이상 절대 넘어갈 일이 없어요.

yebink reviewed (changes requested):
수고했어요 예지니어스! 👍 
꽤 무거운 작업이었군요!

몇 가지 리뷰 남겼으니 확인 부탁해요 🙂

LTTTTTE commented on front/src/components/member/MemberSidebar.vue:
```diff
@@ -53,6 +53,7 @@ const LOGOUT_SUCCESS_MESSAGE = "성공적으로 로그아웃 되었습니다.";
 const FAILED_TO_DELETE_MEMBER_MESSAGE =
   "탈퇴에 실패했습니다. 다시 요청해주세요.";
 const DELETE_MEMBER_SUCCESS = "성공적으로 탈퇴를 완료했습니다.";
+const NICKNAME_LENGTH = 11;
```
```suggestion
const MAX_NICKNAME_LENGTH = 11;
```

LTTTTTE commented on front/src/components/member/MemberUpdateModal.vue:
```diff
@@ -35,12 +38,22 @@
       newNickname: "",
+      rules: {
+        violated: (newVal) =>
+          !!newVal.match(NICKNAME_REGEX) ||
```
```suggestion
          NICKNAME_REGEX.test(newVal) ||
```

느낌표 두개 안쓰는 방법~
(이거 하려면 REGEX는 아래로 수정되어야함)
`/^[a-zA-Zㄱ-ㅎㅏ-ㅣ가-힣0-9]+$/`

chws commented on front/src/components/member/MemberSidebar.vue:
```diff
@@ -83,6 +84,18 @@ export default {
     isGuest() {
       return this.$store.getters["member/isGuest"];
     },
+    shortNickname() {
```
```suggestion
    shortenedNickname() {
```
```suggestion
    shortNicknameForSidebar() {
```
둘 중 하나로 바꾸는게 어떨까요? 처음에 변수명만 봤을 때 이해가 한번에 안가네요😢 
오구가 제안하신대로 `MAX_NICKNAME_LENGTH` 와 함께 적용하면 좋을 것 같아요

chws reviewed (commented):
수고하셨습니다 예지니어스
생각보다 작업이 세세하게 많이 들어갔네요.

하나 논의하고 싶은 사항이, 닉네임의 제한사항(숫자, 영어, 한글만 사용가능)을 백엔드에서는 현재 검사하지 않고 있더라고요. 백엔드에서도 유효성 검사를 해줘야 하지 않나 싶네요... 추가하면 어떨까요? 특수문자가 포함된 닉네임과 같이 유효하지 않은 닉네임을 걸러내야 할 것 같습니다!

YejiAhn committed:
feature/196 refactor: css 관련 수정

 - 오버라이드되는 ma 속성 mx로 수정

YejiAhn committed:
feature/196 refactor: 닉네임 길이제한 관련 코드 리팩토링

 - 최대 닉네임 글자 제한 상수명 MAX_NICKNAME_LENGTH 로 리팩토링
 - 배열로 변환하는 과정 삭제
 - 사이드바에서 잘려서 보여지는 닉네임 변수명 shortenedNickname으로 수정

YejiAhn committed:
feature/196 refactor: 닉네임 관련 제한사항 규칙들을 한꺼번에 묶어서 관리하게 리팩토링

YejiAhn committed:
feature/196 refactor: 변경한 닉네임이 기존 닉네임과 같을 때 메세지 없이 저장버튼만 비활성화되게 리팩토링

YejiAhn committed:
feature/196 refactor: 닉네임 제한사항 관련 Test를 ParameterizedTest로 묶어서 관리

YejiAhn committed:
feature/196 feat: Member 닉네임 정규표현식 관련 제한사항 설정

YejiAhn committed:
feature/196 refactor: 저장 버튼을 비활성화시켜주는 작업을 methods에서 computed로 이동

YejiAhn committed:
feature/196 refactor: 재사용되는 메서드 분리

chws commented on src/main/java/com/grasshouse/dorandoran/member/domain/Member.java:
```diff
@@ -34,6 +36,8 @@
     private Long id;
 
+    @Length(max = 15, message = "닉네임은 15자를 초과할 수 없습니다.")
+    @Pattern(regexp = "^[a-zA-Zㄱ-ㅎㅏ-ㅣ가-힣0-9]+$", message = "닉네임은 한글/영어/숫자로만 이루어져야 합니다.")
```
👍

chws reviewed (approved):
수고하셨습니다~!

hwookim commented on front/src/components/member/MemberUpdateModal.vue:
```diff
@@ -35,18 +38,33 @@
 import api from "@/api/member";
 
+const NICKNAME_REGEX = "^[a-zA-Zㄱ-ㅎㅏ-ㅣ가-힣0-9]+$";
+const VIOLATE_REGEX = (val) => !val.match(NICKNAME_REGEX);
+const IS_VALUE_SAME = (newVal, oldVal) => newVal === oldVal;
```
상수로 보기는 힘든 거 같아요.
메소드로 사용하는 게 더 맞지 않을까요?

```js
  methods: {
    isSame(newVal, oldVal) {
      return newVal === oldVal;
    },
    validateRegex(val) {
      return !val.match(NICKNAME_REGEX);
    },
  ...
```

yebink replied:
저는 상수 형태로 둬도 좋을 것 같아요!
이미 api도 const 형태이고, 저렇게 놔두니 regex와 관련된 부분을 한 번에 파악할 수 있어서 좋네요 👍

hwookim replied:
@YebinK 

이미 자바스크립트에서 `const`는 모두 상수이긴 하지만...
제가 말했던 것은 일단 형태적인 이야기였어요.
우리가 컨벤션적으로 `VIOLATE_REGEX`의 형태를 항상 일정한 값이 있는 진짜 상수로 취급해서잖아요?
마치 `constants.js` 안에 있는 친구들 처럼요.

적어도 `const validateRegex = (val) => ...`로 표현해야하지 않을까요?

이제 여기서 해당 메소드가 외부에서 쓰일 일이 없다면 `methods` 안에 포함되는 게 맞다고 보고요.

hwookim commented on front/src/components/member/MemberUpdateModal.vue:
```diff
@@ -35,18 +38,33 @@
       return this.$store.getters["member/getMember"];
     },
+    isUpdateDisabled() {
+      return (
+        IS_VALUE_SAME(this.member.nickname, this.newNickname) ||
+        VIOLATE_REGEX(this.newNickname)
+      );
```
```suggestion
      return (
        (this.member.nickname === this.newNickname) ||
        VIOLATE_REGEX(this.newNickname)
      );
```

따로 rules에서 관리를 안할 로직이라면 빼둘 필요가 있을까 싶네요.

hwookim reviewed (changes requested):
어... 뭔가 프로젝트 전체적으로 문제가 많이 생긴 듯 하군요...ㅠㅠㅠㅠ
전체적인 교통 정리가 한번 필요할 것 같네요...😢

간단한 리뷰 몇가지 남겼어요. 확인부탁드려요!

yebink commented on src/test/java/com/grasshouse/dorandoran/member/repository/MemberRepositoryTest.java:
```diff
@@ -0,0 +1,44 @@
+import org.springframework.boot.test.context.SpringBootTest;
+
+@SpringBootTest
+public class MemberRepositoryTest {
```
엔티티 validation에 관한 테스트는 어떻게 하나 궁금했는데
repository test를 하는 방법이 있군요 👍

yebink reviewed (approved):
수고했어요 예지니어스 👍 
rules가 도입되다니 너무 기뻐요!

테코톡과 겹쳐 고생이 많았을 것 같네요. 이만 approve합니다~

LTTTTTE reviewed (approved):
고생하셨습니다~
커밋줄수가 1만줄이넘네요 🐰 ㅋㅋㅋ
pull rebase 한번 해주세요~

YejiAhn committed:
feature/196 feat: 상수로 쓰이던 함수 위치를 methods로 변경
